<!DOCTYPE html>
    <html>
    <head>
        <meta charset="UTF-8">
        <title>Functional annotation of rapidly evolving orthogroups</title>
        <style>
/* From extension vscode.github */
/*---------------------------------------------------------------------------------------------
 *  Copyright (c) Microsoft Corporation. All rights reserved.
 *  Licensed under the MIT License. See License.txt in the project root for license information.
 *--------------------------------------------------------------------------------------------*/

.vscode-dark img[src$=\#gh-light-mode-only],
.vscode-light img[src$=\#gh-dark-mode-only] {
	display: none;
}

/* From extension ms-toolsai.jupyter */
/* These classnames are inherited from bootstrap, but are present in most notebook renderers */

.alert {
    width: auto;
    padding: 1em;
    margin-top: 1em;
    margin-bottom: 1em;
}
.alert > *:last-child {
    margin-bottom: 0;
}
#preview > .alert:last-child {
    /* Prevent this being set to zero by the default notebook stylesheet */
    padding-bottom: 1em;
}

.alert-success {
    /* Note there is no suitable color available, so we just copy "info" */
    background-color: var(--theme-info-background);
    color: var(--theme-info-foreground);
}
.alert-info {
    background-color: var(--theme-info-background);
    color: var(--theme-info-foreground);
}
.alert-warning {
    background-color: var(--theme-warning-background);
    color: var(--theme-warning-foreground);
}
.alert-danger {
    background-color: var(--theme-error-background);
    color: var(--theme-error-foreground);
}

</style>
        
        <link rel="stylesheet" href="https://cdn.jsdelivr.net/gh/Microsoft/vscode/extensions/markdown-language-features/media/markdown.css">
<link rel="stylesheet" href="https://cdn.jsdelivr.net/gh/Microsoft/vscode/extensions/markdown-language-features/media/highlight.css">
<style>
            body {
                font-family: -apple-system, BlinkMacSystemFont, 'Segoe WPC', 'Segoe UI', system-ui, 'Ubuntu', 'Droid Sans', sans-serif;
                font-size: 14px;
                line-height: 1.6;
            }
        </style>
        <style>
.task-list-item {
    list-style-type: none;
}

.task-list-item-checkbox {
    margin-left: -20px;
    vertical-align: middle;
    pointer-events: none;
}
</style>
        
    </head>
    <body class="vscode-body vscode-light">
        <h1 id="functional-annotation-of-rapidly-evolving-orthogroups">Functional annotation of rapidly evolving orthogroups</h1>
<p>The goal here is to identify interesting functional annotation orthogroups that can be summarized in a table in the main text. Probably like 8-15 or so. Ones that i definitely want to mention:</p>
<ul>
<li>Whatever is going on with the giant ones in A. obtectus
<ul>
<li>Transposases misannotated in flybase</li>
</ul>
</li>
<li>detoxification stuff about cytochrome P450
<ul>
<li>some interesting stuff potentially about polyethylene breakdown in tenebrioids?</li>
</ul>
</li>
<li>Some stuff about the genes annotated with sexual reproduction
<ul>
<li>Nothing concrete</li>
</ul>
</li>
<li>Some stuff about the olfactory genes
<ul>
<li>the chitin stuff is development, not odorant binding</li>
</ul>
</li>
<li>The fluorescence stuff in i. luminosius and the other glowing one
<ul>
<li>esterase and luciferase</li>
</ul>
</li>
<li>Growth factor and development
<ul>
<li>the chitin stuff that is not pheromone sensing</li>
</ul>
</li>
</ul>
<p>Other considerations:</p>
<ul>
<li>T. molitor and Z. morio are of particular interest because they are of industrial relevance, both as feed insects for agriculture and exotic pets, and as human food (alternative to other animal proteins). (They also use BRAKER2 and then TSEBRA manually with the default weights!! they also get high gene numbers and lots of single exon genes, especially z. morio)
<ul>
<li>T. molitor RNA: SRR18735292</li>
<li>z. morio RNA: SRR18735291</li>
</ul>
</li>
</ul>
<p>I am doing stuff with the table in <code>/PhD_chapter1/src/functional_annotation_eval.py</code>.</p>
<h2 id="cafe-significant-orthogroups">CAFE significant orthogroups</h2>
<p>CAFE is a maximum likelihood method and needs to be run multiple times to ensure convergence. I have run CAFE 20 times, and the lambda values all converge nicely to ca. 0.4067, and each run has between 530 and 570 significant orthogorups. I want to inlcude only orthogroups where I can be certain they are significant, so I decided to only include ones significant in all runs, which are 496 orthogroups (out of 8315 considered.)</p>
<p><img src="file:////Users/miltr339/work/PhD_code/PhD_chapter1/data/CAFE_convergence/runs_sig_OGs_hist.svg" alt="Sig OG abundance histogram" width=30%></p>
<p>Other pre-processing I did to the CAFE input:</p>
<ol>
<li>only include orthogroups that have at least one member in <em>D. melanogaster</em> because this means they are for sure only present at the root of the tree.</li>
<li>only inlcude orthogroups with fewer than 100 gene family members in any species. This excludes (after the first step) only two more, which have huge expansions in <em>A. obtectus</em> and are explained below.</li>
</ol>
<h2 id="large-expansion-in-a-obtectus">large expansion in <em>A. obtectus</em></h2>
<p>A. obtectus has two large gene families, N0.HOG0000035 and N0.HOG0000014, which are twice as large as the third largest one. The flybase IDs of these orthogroups have no functional annotation, and are not associated with any DAVID Gene Family Cluster. The correlation with genome size and repeat content is <em>significant</em> in both cases. Considering that most of the other orthogroups investigated here are not significant in their association, maybe that means this is not a biological signal for selection based on function, but instead the expansion is caused by drift through TE activity (and whatever the genome size does)?</p>
<p>I used the <a href="https://www.ncbi.nlm.nih.gov/Structure/cdd/wrpsb.cgi">ncbi conserved domain search</a> on all the proteins in these orthogroups, and for both of them most proteins match transposases, and sometimes at the c-terminus zinc-finger related domains (which is also common for transposases), which probably means that this is a false positive gene. I also then checked the uniform <em>D. melanogaster</em> orthologs and the corresponding flybase IDs (FBgn0037633 for N0.HOG0000035 and FBgn0033750 for N0.HOG0000014) and all of those also follow the pattern of a very good transposase hit and a less good zinc-finger hit at the C-terminus. Therefore this is likely a false positive.</p>
<p><img src="file:////Users/miltr339/work/PhD_code/PhD_chapter1/data/functional_annot_eval/Aobt_expansion_GF_sizes.svg" alt="A. obtectus expansion" width=45%></p>
<h1 id="todo-re-do-pictures-and-double-check-all-from-here-on-down">TODO re-do pictures and double check all from here on down</h1>
<h2 id="detoxification">Detoxification</h2>
<p>Gene Family Cluster 1 is the main detoxification group, with all the cytochrome P450 in it, but Gene Family Cluster 17 which has aldehyde deoxide in it also has something to do with detoxification. All of these orthogroups are very quiet in <em>Bruchinae</em> and <em>Curculionidae</em> (<em>D. ponderosae</em> and <em>R. ferrugineus</em>). Notable expansions are especially <em>Elateriforma</em> (<em>I. luminosus</em>) which are fluorescent, and <em>Tenebrioids</em>. The two functional anotations in this category are:</p>
<ul>
<li>cytochrome P450 (CYP) is detoxification and we think it could be related to host plant adaptation?</li>
<li>Aldehyde oxidase is a part of the acetaldehyde metabolic process, which breaks down acetaldehyde into less toxic substances</li>
</ul>
<p><em>Tenebrioid</em> insects such as <em>T. molitor</em> and <em>Z. morio</em> can break down plasctics in their gut as larvae. This is possible through specific microorganisms in their gut working together with insect-encoded proteins that are secreted (the degradation rate of isolated microorganisms is significantly lower). <a href="https://link.springer.com/article/10.1007/s10924-023-03029-z#Sec11">Source</a> for all that follows. The first step is polyethylene being oxidated by Cytochrome p450 (and others), then more chemistry (involving aldehyde dehydrogenase, which is not the same as aldehyde oxidase!), until it is broken down by lipid metabolism in the cells. According to the flybase annotations, the expression profiles of the Gene Family Cluster 1 orthogroups have their peak mostly during larval, and pupae stages, but not exclusively (some are adult sex specific). The differential expression analysis does not provide super strong evidence for CYP though.</p>
<h3 id="gene-family-cluster-1-cytochrome-p450">Gene Family Cluster 1: Cytochrome P450</h3>
<p>All Orthogroups in this Gene Family Cluster are very insignificant for the correlation with both repeat content and genome size. I am wondering if this is a good thing, since the other functional stuff below is not significantly correlated, and e.g. the false-positive annotations for the large orthogroups in <em>A. obtectus</em> are.</p>
<p>This proliferation is unsurprising (expected even?), especially in insects (<a href="https://link.springer.com/article/10.1186/1471-2164-14-174">source</a>, <em>T. castaneum</em> already has more duplications in the CYP superfamily than <em>B. mori</em> or <em>D. melanogaster</em>). Also detoxification, yes, but they also talk about it in relation to larval development and egg maturation. Some subfamilies (existing but not found as significant here) have also been linked to insecticide resistance in <em>T. castaneum</em>. Cytochrome P450 according to <a href="https://link.springer.com/article/10.1186/1471-2164-14-174#Sec2">this paper</a> is also important in mitochondria, but Gene Family Cluster 33, which is mitochondrial translation shows no interesting dynamics.</p>
<h3 id="gene-family-cluster-3-lipid-metabolic-process">Gene Family Cluster 3: lipid metabolic process</h3>
<p>Since the lipid metabolism in the cells is the end of the polyethylene breakdown, this Gene Family Cluster is also relevant, because it's the last step in the digestion process in the cells. N0.HOG0000085 shows an expansion in <em>Z. morio</em>, which also relates to the polyethylene synthesis described above. The other two are pretty boring</p>
<h3 id="gene-family-cluster-18-aldehyde-oxidase">Gene Family Cluster 18: aldehyde oxidase</h3>
<p>This Gene Family Cluster contains only aldehyde oxidase. The only rapidly evolving orthogroup is N0.HOG0000669, which is insignificant for the correlation with repeat content and also with genome size. It is expressed during all developmental stages and adults of both sexes. There is what seems like a significant expansion in <em>I. luminosus</em>, but <em>I. luminosus</em> has a shit assembly so take that with a grain of salt.</p>
<h3 id="plot">Plot</h3>
<p>
<img src="file:////Users/miltr339/work/PhD_code/PhD_chapter1/data/functional_annot_eval/detoxificatoin_clusters.svg" alt="Gene Family Cluster 17" width=45%>
</p>
<p>Aldehyde is a byproduct of alcohol metabolism (in humans?) and is broken down into less toxic molecules by aldehyde oxidase. In insects, aldehyde can come from other sources, such as pheromones, and &quot;<em>Long-chain, unsaturated alcohol and aldehyde compounds are common  female-produced  sex  pheromones (REF)</em>&quot; (<a href="https://journals.plos.org/plosone/article?id=10.1371/journal.pone.0067794">source</a>). They conclude that &quot;<em>[The] results suggest that an antennae-specific aldehyde oxidasefrom  the  navel  orangeworm,  AtraAOX2,  might  be  involved  indegradation  of  host  plant  volatile  compounds  and  pheromone</em>&quot;, which helps both with distinguishing plant compounds from the pheromones they are actually interested in, and also protects them from volatile plant compounts and potentially even pesticides.</p>
<p>Quote from the abstract: &quot;<em>Our data suggest AtraAOX2 might be involved in degradation of a diversity of aldehydes including sex pheromones, plant-derived semiochemicals, and chemical cues for oviposition sites. Additionally, AtraAOX2 could protect the insect's olfactory system from xenobiotics, including pesticides that might reach the sensillar lymph surrounding the olfactory receptor neurons.</em>&quot; (I don't understand the methods or biochemistry in general enough to judge how accurate this is).</p>
<h2 id="sexual-reproduction">Sexual reproduction</h2>
<p>Here, the relevant Gene Family Clusters are these, and all orthogroups are insiginificant for correlations with GS or repeat content. It seems like immunity and sexual reproduction are tightly linked.</p>
<ul>
<li>Gene Family Cluster 5, Function: protease inhibitor (immunity, reproduction)</li>
<li>Gene Family Cluster 8, Function: immunity and sexual reproduction</li>
<li>Gene Family Cluster 9, Function: sexual reproduction</li>
</ul>
<p>Similar to the detoxification, which means not a lot going on in Bruchinae and Curculionidae, but there's one massively expanding in <em>Z. morio</em> (N0.HOG0000401), whose expression peak is in adult males, and is predicted to be in the ER according to flybase, the molecular function is just protein homodimerization. Otherwise there is not a lot of info, even in <em>D. melanogaster</em>. I have a hard time finding good papers for this based on this extremely vague information.  The other orthogroups are also not super helpful, none show very interesting dynamics that I could read into.</p>
<p>
<img src="file:////Users/miltr339/work/PhD_code/PhD_chapter1/data/functional_annot_eval/sexual_reproduction.svg" alt="Gene Family Cluster 5,8,9" width=45%>
</p>
<h2 id="pheromone-sensing">Pheromone sensing</h2>
<p>This also involves cuticular proteins due to how insects actually do the sensing, so here I think these are the relevant Gene Family Clusters:</p>
<ul>
<li>Gene Family Cluster 7, Function: odorant binding</li>
<li>Gene Family Cluster 20, Function: transmembrane transport (olfactory)</li>
<li>Gene Family Cluster 30, Function: pheromone sensing</li>
</ul>
<p>Chitin can also be related to pheromone sensing due to cuticular hydrocarbons. No orthogroup that is significant in any CAFE run is annotated with cuticular hydrocarbons, and the others with chitin/cuticular functions are expresssed only in early development and regulate chitin formation.</p>
<h3 id="gene-family-cluster-7-20-and-30">Gene Family Cluster 7, 20 and 30</h3>
<p>None of them have a significant correlation with GS or TE content except N0.HOG0001445 in cluster 30.
Cluster 20, ATPase-coupled transmembrane transporter mostly is not annotated on flybase (N0.HOG0000761 is annotated with &quot;response to toxic substance&quot;). According to <a href="https://www.mdpi.com/2075-4450/15/12/1016">this paper</a>, it's expressed in antennae in drosophila and might be related to odorant processing.</p>
<p>
<img src="file:////Users/miltr339/work/PhD_code/PhD_chapter1/data/functional_annot_eval/pheromone_sensing_clusters.svg" alt="Pheromone sensing" width=45%>
</p>
<p>The highlight is N0.HOG0000037 for Gene Family Cluster 30 and N0.HOG0000056 for Gene Family Cluster 7, which make the &quot;M&quot; shape with the peaks at <em>A. verrucosus</em> and <em>Z. morio</em> or <em>T. molitor</em> respectively. Both of them have their expression peak in adult males.</p>
<p>N0.HOG0000038 (GG30) and N0.HOG0000436 (GG7) have the high peak in <em>I. luminosus</em>.</p>
<h2 id="development">Development</h2>
<h3 id="gene-family-cluster-15-24-and-26">Gene Family Cluster 15, 24, and 26</h3>
<p>Chitin related and cuticular protein. No significant correlation with repeats or GS not even in the large cluster 26, except N0.HOG0000108 in Cluster 24</p>
<p>There is again expansions in <em>Z. morio</em>, for both categories. for GC16, it's N0.HOG0001194 and N0.HOG0003035, the former (higher peak) is only expressed in early embryonic development, but the latter is an intercellular matrix component also expressed in adults, so it might have something to do with odorants. All of Gene Family Cluster 23 is stuff only expressed during larval development and is responsible for cuticle development, nothing to do with odorant receptors. For this reason I decided to investigate other Gene Family Clusters that also have function in early development, such as Gene Family cluster 26 and 13. TODO</p>
<p>Also this time there's stuff going on in bruchids! GG23:N0.HOG0000108 GG16:N0.HOG0000307 have expansions in bruchids. However, both of these are involved in cuticular development during larval and pupae stages, where their expression peak also happens, so this is not related to adult pheromone sensing.</p>
<p>
<img src="file:////Users/miltr339/work/PhD_code/PhD_chapter1/data/functional_annot_eval/Gene_Group_16_chitin_related_GF_sizes.svg" alt="Gene Family Cluster 16" width=45%>
<img src="file:////Users/miltr339/work/PhD_code/PhD_chapter1/data/functional_annot_eval/Gene_Group_23_cuticular_protein_GF_sizes.svg" alt="Gene Family Cluster 23" width=45%>
</p>
<h3 id="new-adenosine-deaminase-related-growth-factor">new: Adenosine deaminase-related growth factor</h3>
<p>Recently identified growth factors: <a href="https://www.sciencedirect.com/science/article/abs/pii/S0378111901007624">paper</a></p>
<h2 id="fluorescence-in-elateriforma-">Fluorescence in <em>Elateriforma</em> (?)</h2>
<h3 id="gene-family-cluster-4">Gene Family Cluster 4</h3>
<p>Gene Family Cluster 4 (Esterase and mating behavior) expanded in <em>Elateriforma</em>, and is also common in the significantly rapidly expanding gene families, which is not the case in other species. It also expands in <em>Z. morio</em>, but there only 3 of the expanding gene families overall are in this group, and this species has a lot of expansions in general</p>
<p>
<img src="file:////Users/miltr339/work/PhD_code/PhD_chapter1/data/functional_annot_eval/Gene_Group_4_esterase_GF_sizes.svg" alt="Gene Family Cluster 4" width=45%>
</p>
<h3 id="acyl-coa-synthetase">Acyl-CoA synthetase</h3>
<p>According to <a href="https://web.archive.org/web/20180722105054id_/https://www.biorxiv.org/content/biorxiv/early/2017/12/21/237586.full.pdf">this paper</a> (especially figure 3), the <em>Elateriforma</em> share a common genomic mechanism for their fluorescence made up of luciferases and their paralogs, which form a cluster of genes (that's what they call it, no clue what that means exactly. I assume a gene family since they talk about how it came to be through tandem duplication later) containing the luciferase and close relatives such as peroxisomal fatty acyl-CoA synthetase (PACS) and non-peroxisomal acyl-CoA synthetase (ACS). This is not represented in any Gene Family Cluster, but the gene family N0.HOG0000613 is annotated as <em>Acyl-CoA synthetase family member 2</em> (but there's no other family members annotated). Some other orthogroups have Acyl-CoA synthesis-related annotations (not oxidase/reductase/hydrolase!) in their API summaries:</p>
<ul>
<li>N0.HOG0000284 (maybe a good candidate for the table?)</li>
<li>N0.HOG0000397 (pudgy in Dmel)</li>
</ul>
<p>Seems like these all show expansions in <em>Elateriforma</em>, cool!</p>
<p>
<img src="file:////Users/miltr339/work/PhD_code/PhD_chapter1/data/functional_annot_eval/N0_HOG0000613_GF_sizes.svg" alt="Gene Family Cluster 4" width=45%>
<img src="file:////Users/miltr339/work/PhD_code/PhD_chapter1/data/functional_annot_eval/Acyl_CoA_synthesis.svg" alt="Acyl_CoA_synthesis" width=45%>
</p>
<h1 id="general-enrichment-of-gene-family-clusters-in-rapidly-expanding-gene-families">General &quot;enrichment&quot; of Gene Family Clusters in rapidly expanding gene families</h1>
<p>I looked at all the gene families in a species, and selected the ones that are in the upper 5th size percentile of all gene families (in species with few genes, this is sometimes just with one or more members, so I also set a min. size limit of 2 members here). Then I checked which ones are significantly rapidly evolving according to CAFE, and of these which ones have been assigned to functional Gene Family Clusters by their <em>D. melanogaster</em> member.  This is only a very small subset of all gene families in the upper 5th size percentile. I have then compiled a file that shows how many of the above selected gene families fall under each functional Gene Family Cluster. I think this can get a more general idea of the kind of things that might be evolving in some of the species. The full file is here: <code>PhD_chapter1/data/functional_annot_eval/frequent_Gene_Groups_in_expanding_GFs.txt</code> and I am going to show the output for all species below, only including Gene Family Clusters that appear at least four times and are not uncharacterized or unannotated.</p>
<p>These are my conclusions:</p>
<ul>
<li>Gene Family Cluster 25 (glycolysis and early development), and Gene Family Cluster 1 (detoxification, cytochrome P450) are basically in all of them.</li>
<li>Gene Family Cluster 2 (proteolysis) is in all families except bruchids (where it only occurs in B. siliquastri)</li>
<li>Tenebrioids all have Gene Family Cluster 30 (pheromone sensing)</li>
<li>Elateriformia all have Gene Family Cluster 4: Esterase and mating behavior.</li>
</ul>
<h2 id="shortened-output-of-functional-summary">Shortened output of functional summary</h2>
<h4 id="d_melanogaster">D_melanogaster</h4>
<p>number of gene families with more than 2 members (upper 5th percentile) = 506 (of 8760 orthogroups). 25 unique Gene Family Clusters with functional annotations, these ones appear at least four times:</p>
<ul>
<li>11 GFs annotated as : glycolysis and early development (and other)</li>
<li>10 GFs annotated as : protein breakdown (proteolysis)</li>
<li>6 GFs annotated as : detoxification</li>
</ul>
<h3 id="elateriformia">Elateriformia</h3>
<h4 id="i_luminosus">I_luminosus</h4>
<p>number of gene families with more than 3 members (upper 5th percentile) = 1276 (of 13090 orthogroups). 27 unique Gene Family Clusters with functional annotations, these ones appear at least four times:</p>
<ul>
<li>15 GFs annotated as : protein breakdown (proteolysis)</li>
<li>11 GFs annotated as : glycolysis and early development (and other)</li>
<li>6 GFs annotated as : detoxification</li>
<li>4 GFs annotated as : Esterase and mating behavior</li>
</ul>
<h4 id="p_pyralis">P_pyralis</h4>
<p>number of gene families with more than 3 members (upper 5th percentile) = 1213 (of 12644 orthogroups). 27 unique Gene Family Clusters with functional annotations, these ones appear at least four times:</p>
<ul>
<li>20 GFs annotated as : glycolysis and early development (and other)</li>
<li>11 GFs annotated as : protein breakdown (proteolysis)</li>
<li>6 GFs annotated as : detoxification</li>
<li>5 GFs annotated as : transcription regulation</li>
<li>4 GFs annotated as : odorant binding</li>
<li>4 GFs annotated as : Esterase and mating behavior</li>
</ul>
<h3 id="coccinellidae">Coccinellidae</h3>
<h4 id="c_septempunctata">C_septempunctata</h4>
<p>number of gene families with more than 2 members (upper 5th percentile) = 968 (of 10892 orthogroups). 27 unique Gene Family Clusters with functional annotations, these ones appear at least four times:</p>
<ul>
<li>16 GFs annotated as : glycolysis and early development (and other)</li>
<li>5 GFs annotated as : protein breakdown (proteolysis)</li>
<li>4 GFs annotated as : detoxification</li>
</ul>
<h3 id="tenebrionidae">Tenebrionidae</h3>
<h4 id="a_verrucosus">A_verrucosus</h4>
<p>number of gene families with more than 2 members (upper 5th percentile) = 914 (of 12405 orthogroups). 27 unique Gene Family Clusters with functional annotations, these ones appear at least four times:</p>
<ul>
<li>28 GFs annotated as : glycolysis and early development (and other)</li>
<li>10 GFs annotated as : protein breakdown (proteolysis)</li>
<li>6 GFs annotated as : detoxification</li>
<li>6 GFs annotated as : pheromone sensing</li>
</ul>
<h4 id="t_castaneum">T_castaneum</h4>
<p>number of gene families with more than 2 members (upper 5th percentile) = 548 (of 11778 orthogroups). 31 unique Gene Family Clusters with functional annotations, these ones appear at least four times:</p>
<ul>
<li>27 GFs annotated as : glycolysis and early development (and other)</li>
<li>8 GFs annotated as : protein breakdown (proteolysis)</li>
<li>7 GFs annotated as : detoxification</li>
<li>6 GFs annotated as : pheromone sensing</li>
</ul>
<h4 id="z_morio">Z_morio</h4>
<p>number of gene families with more than 2 members (upper 5th percentile) = 1373 (of 13008 orthogroups). 34 unique Gene Family Clusters with functional annotations, these ones appear at least four times:</p>
<ul>
<li>28 GFs annotated as : glycolysis and early development (and other)</li>
<li>12 GFs annotated as : protein breakdown (proteolysis)</li>
<li>6 GFs annotated as : pheromone sensing</li>
<li>5 GFs annotated as : detoxification</li>
<li>4 GFs annotated as : chromatin organization and transcription regulation</li>
<li>4 GFs annotated as : neurological (mostly uncharacterized)</li>
</ul>
<h4 id="t_molitor">T_molitor</h4>
<p>number of gene families with more than 2 members (upper 5th percentile) = 732 (of 11174 orthogroups). 34 unique Gene Family Clusters with functional annotations, these ones appear at least four times:</p>
<ul>
<li>19 GFs annotated as : glycolysis and early development (and other)</li>
<li>8 GFs annotated as : protein breakdown (proteolysis)</li>
<li>7 GFs annotated as : detoxification</li>
<li>6 GFs annotated as : pheromone sensing</li>
</ul>
<h3 id="curculionidae">Curculionidae</h3>
<h4 id="d_ponderosae">D_ponderosae</h4>
<p>number of gene families with more than 2 members (upper 5th percentile) = 839 (of 10658 orthogroups): 23 unique Gene Family Clusters with functional annotations, these ones appear at least four times:</p>
<ul>
<li>24 GFs annotated as : glycolysis and early development (and other)</li>
<li>10 GFs annotated as : protein breakdown (proteolysis)</li>
<li>5 GFs annotated as : detoxification</li>
</ul>
<h4 id="r_ferrugineus">R_ferrugineus</h4>
<p>number of gene families with more than 2 members (upper 5th percentile) = 600 (of 11320 orthogroups). 23 unique Gene Family Clusters with functional annotations, these ones appear at least four times:</p>
<ul>
<li>19 GFs annotated as : glycolysis and early development (and other)</li>
<li>7 GFs annotated as : protein breakdown (proteolysis)</li>
<li>6 GFs annotated as : detoxification</li>
</ul>
<h3 id="bruchinae">Bruchinae</h3>
<h4 id="a_obtectus">A_obtectus</h4>
<p>number of gene families with more than 3 members (upper 5th percentile) = 1211 (of 12608 orthogroups). 20 unique Gene Family Clusters with functional annotations, these ones appear at least four times:</p>
<ul>
<li>16 GFs annotated as : glycolysis and early development (and other)</li>
<li>12 GFs annotated as : transcription regulation</li>
<li>7 GFs annotated as : detoxification</li>
<li>5 GFs annotated as : chromatin organization and transcription regulation</li>
</ul>
<h4 id="b_siliquastri">B_siliquastri</h4>
<p>number of gene families with more than 2 members (upper 5th percentile) = 433 (of 11255 orthogroups). 18 unique Gene Family Clusters with functional annotations, these ones appear at least four times:</p>
<ul>
<li>13 GFs annotated as : glycolysis and early development (and other)</li>
<li>5 GFs annotated as : detoxification</li>
<li>4 GFs annotated as : protein breakdown (proteolysis)</li>
</ul>
<h4 id="c_chinensis">C_chinensis</h4>
<p>number of gene families with more than 3 members (upper 5th percentile) = 877 (of 12655 orthogroups). 25 unique Gene Family Clusters with functional annotations, these ones appear at least four times:</p>
<ul>
<li>18 GFs annotated as : glycolysis and early development (and other)</li>
<li>6 GFs annotated as : chromatin organization and transcription regulation</li>
<li>5 GFs annotated as : transcription regulation</li>
<li>4 GFs annotated as : detoxification</li>
</ul>
<h4 id="c_maculatus">C_maculatus</h4>
<p>number of gene families with more than 4 members (upper 5th percentile) = 1106 (of 14402 orthogroups). 22 unique Gene Family Clusters with functional annotations, these ones appear at least four times:</p>
<ul>
<li>9 GFs annotated as : glycolysis and early development (and other)</li>
<li>5 GFs annotated as : detoxification</li>
</ul>
<h1 id="papers">Papers</h1>
<p>These are papers I didn't cite anywhere above but might still want to reference in my manuscript.</p>
<ul>
<li>tenebrioid assembly and gene family evolution: <a href="https://academic.oup.com/g3journal/article/13/6/jkad079/7099445">https://academic.oup.com/g3journal/article/13/6/jkad079/7099445</a></li>
<li>beetles are super underrepresented in sequencing projects: <a href="https://doi.org/10.1073/pnas.2109019118">https://doi.org/10.1073/pnas.2109019118</a></li>
</ul>
<h1 id="table">TABLE</h1>
<p>These are the orthogroups that should be included in a table in the paper</p>
<ul>
<li>N0.HOG0000035</li>
<li>N0.HOG0000014</li>
<li>N0.HOG0000140</li>
<li>N0.HOG0000085</li>
<li>N0.HOG0000401</li>
<li>N0.HOG0000037</li>
<li>N0.HOG0000056</li>
<li>N0.HOG0000284</li>
</ul>

        
        
    </body>
    </html>