<!DOCTYPE html>
    <html>
    <head>
        <meta charset="UTF-8">
        <title>Correlations of gene family sizes vs&period; species genome characteristics</title>
        <style>
/* From extension vscode.github */
/*---------------------------------------------------------------------------------------------
 *  Copyright (c) Microsoft Corporation. All rights reserved.
 *  Licensed under the MIT License. See License.txt in the project root for license information.
 *--------------------------------------------------------------------------------------------*/

.vscode-dark img[src$=\#gh-light-mode-only],
.vscode-light img[src$=\#gh-dark-mode-only] {
	display: none;
}

/* From extension ms-toolsai.jupyter */
/* These classnames are inherited from bootstrap, but are present in most notebook renderers */

.alert {
    width: auto;
    padding: 1em;
    margin-top: 1em;
    margin-bottom: 1em;
}
.alert > *:last-child {
    margin-bottom: 0;
}
#preview > .alert:last-child {
    /* Prevent this being set to zero by the default notebook stylesheet */
    padding-bottom: 1em;
}

.alert-success {
    /* Note there is no suitable color available, so we just copy "info" */
    background-color: var(--theme-info-background);
    color: var(--theme-info-foreground);
}
.alert-info {
    background-color: var(--theme-info-background);
    color: var(--theme-info-foreground);
}
.alert-warning {
    background-color: var(--theme-warning-background);
    color: var(--theme-warning-foreground);
}
.alert-danger {
    background-color: var(--theme-error-background);
    color: var(--theme-error-foreground);
}

</style>
        
        <link rel="stylesheet" href="https://cdn.jsdelivr.net/gh/Microsoft/vscode/extensions/markdown-language-features/media/markdown.css">
<link rel="stylesheet" href="https://cdn.jsdelivr.net/gh/Microsoft/vscode/extensions/markdown-language-features/media/highlight.css">
<style>
            body {
                font-family: -apple-system, BlinkMacSystemFont, 'Segoe WPC', 'Segoe UI', system-ui, 'Ubuntu', 'Droid Sans', sans-serif;
                font-size: 14px;
                line-height: 1.6;
            }
        </style>
        <style>
.task-list-item {
    list-style-type: none;
}

.task-list-item-checkbox {
    margin-left: -20px;
    vertical-align: middle;
    pointer-events: none;
}
</style>
        
    </head>
    <body class="vscode-body vscode-light">
        <h1 id="correlations-of-gene-family-sizes-vs-species-genome-characteristics">Correlations of gene family sizes vs. species genome characteristics</h1>
<p>We hypothesize that gene families proliferate more in genomes with a high repeat content, because some TEs can facilitate gene duplications. Therefore we calculate the linear regression in each significantly rapidly evolving orthogroup between gene family size and several genome characteristics such as genome size, overall repeat content, and repeat-category content.</p>
<h2 id="method">Method</h2>
<h3 id="phylogeneticlaly-independent-contrasts-pics">Phylogeneticlaly independent contrasts (PICs)</h3>
<p>I am using my implementation of the phylogenetically independent contrasts according to Felsenstein 1985 and the species tree generated by orthofinder, transformed to be ultrametric with the ape package in R. The code is here: <code>PhD_chapter1/src/compute_PIC.py</code>. Both explanatory and response variables are transformed.</p>
<h3 id="linear-regression">linear regression</h3>
<p>I am calculating the linear regression with <code>scipy.stats.linregress(genome_characteristic, GF_sizes)</code>.</p>
<ul>
<li>exclude <em>D. melanogaster</em> from the analysis</li>
<li>Both variables are PICs log2-transformed to make them more linear.</li>
</ul>
<p>Unfortunately the sampling is pretty sparse and unequally distributed in the parameter space, and also the gene family sizes are discrete, so a lot of them don't have normally distributed residuals which makes the p-value of the regression unreliable. They often also are not suitable for non-linear regression since they also don't really follow an exponential trend or anything else that could be modelled, and the sample size is really low (13 species). Therefore I have excluded these orthogroups</p>
<h3 id="genome-characteristics">genome characteristics</h3>
<p>I test the gene family sizes in each orthogroup against several genome characteristics to find orthogroups that have significant correlations. Each group is FDR-corrected against multiple testing. The categories are these:</p>
<ul>
<li>Genome size (assembly size)</li>
<li>Overall genome repeat proportion</li>
<li>proportion of different repeat categories</li>
</ul>
<h2 id="results">Results</h2>
<p>There are 496 significantly rapidly evolving orthogroups in total, but each plot shows fewer because I excluded the ones with non-normal residuals as explained above. Each plot is the orthogroup size vs. the linear regression slope to also take into account that larger orthogroups might be more prone to correlations with repetitive elements.</p>
<p>No category shows significantly correlated orthogroups after multiple testing correction.</p>
<h3 id="genome-size">Genome size</h3>
<p><img src="file:////Users/miltr339/work/PhD_code/PhD_chapter1/data/correlations/orthoDB_sig_OGs_vs_GS_inclines_bh_corrected_PIC_vs_OG_size_sig_OGs_colors.svg" alt="correlation" width=30%></p>
<h3 id="repeat-content">Repeat content</h3>
<p>This is the general genome-wide repeat content with no consideration for how it is divided in the different repeat categories</p>
<p><img src="file:////Users/miltr339/work/PhD_code/PhD_chapter1/data/correlations/orthoDB_sig_OGs_vs_reps_inclines_bh_corrected_PIC_vs_OG_size_sig_OGs_colors.png" alt="correlation" width=30%></p>
<h4 id="repeat-categories">repeat categories</h4>
<p><img src="file:////Users/miltr339/work/PhD_code/PhD_chapter1/data/correlations/orthoDB_sig_OGs_vs_Unclassified_inclines_bh_corrected_PIC_vs_OG_size_sig_OGs_colors.svg" alt="correlation" width=29%>
    <img src="file:////Users/miltr339/work/PhD_code/PhD_chapter1/data/correlations/orthoDB_sig_OGs_vs_Lowcomplexity_inclines_bh_corrected_PIC_vs_OG_size_sig_OGs_colors.svg" alt="correlation" width=30%>
    <img src="file:////Users/miltr339/work/PhD_code/PhD_chapter1/data/correlations/orthoDB_sig_OGs_vs_Simplerepeats_inclines_bh_corrected_PIC_vs_OG_size_sig_OGs_colors.svg" alt="correlation" width=30%></p>
<p><img src="file:////Users/miltr339/work/PhD_code/PhD_chapter1/data/correlations/orthoDB_sig_OGs_vs_Rolling-circles_inclines_bh_corrected_PIC_vs_OG_size_sig_OGs_colors.svg" alt="correlation" width=30%>
    <img src="file:////Users/miltr339/work/PhD_code/PhD_chapter1/data/correlations/orthoDB_sig_OGs_vs_Retroelements_inclines_bh_corrected_PIC_vs_OG_size_sig_OGs_colors.svg" alt="correlation" width=30%>
    <img src="file:////Users/miltr339/work/PhD_code/PhD_chapter1/data/correlations/orthoDB_sig_OGs_vs_DNAtransposons_inclines_bh_corrected_PIC_vs_OG_size_sig_OGs_colors.svg" alt="correlation" width=30%></p>
<p>These two repeat categories have the situation where two sister species have a percentage of 0, so the PICs have a zero in them, and the log-transformation for the explanatory variable does not work.</p>
<p><img src="file:////Users/miltr339/work/PhD_code/PhD_chapter1/data/correlations/orthoDB_sig_OGs_vs_SmallRNA_inclines_bh_corrected_PIC_vs_OG_size_sig_OGs_colors.svg" alt="correlation" width=30%>
    <img src="file:////Users/miltr339/work/PhD_code/PhD_chapter1/data/correlations/orthoDB_sig_OGs_vs_Satellites_inclines_bh_corrected_PIC_vs_OG_size_sig_OGs_colors.svg" alt="correlation" width=30%></p>
        
        
    </body>
    </html>