<!DOCTYPE html>
    <html>
    <head>
        <meta charset="UTF-8">
        <title>Correlations of gene family sizes vs&period; species genome characteristics</title>
        <style>
/* From extension vscode.github */
/*---------------------------------------------------------------------------------------------
 *  Copyright (c) Microsoft Corporation. All rights reserved.
 *  Licensed under the MIT License. See License.txt in the project root for license information.
 *--------------------------------------------------------------------------------------------*/

.vscode-dark img[src$=\#gh-light-mode-only],
.vscode-light img[src$=\#gh-dark-mode-only] {
	display: none;
}

/* From extension ms-toolsai.jupyter */
/* These classnames are inherited from bootstrap, but are present in most notebook renderers */

.alert {
    width: auto;
    padding: 1em;
    margin-top: 1em;
    margin-bottom: 1em;
}
.alert > *:last-child {
    margin-bottom: 0;
}
#preview > .alert:last-child {
    /* Prevent this being set to zero by the default notebook stylesheet */
    padding-bottom: 1em;
}

.alert-success {
    /* Note there is no suitable color available, so we just copy "info" */
    background-color: var(--theme-info-background);
    color: var(--theme-info-foreground);
}
.alert-info {
    background-color: var(--theme-info-background);
    color: var(--theme-info-foreground);
}
.alert-warning {
    background-color: var(--theme-warning-background);
    color: var(--theme-warning-foreground);
}
.alert-danger {
    background-color: var(--theme-error-background);
    color: var(--theme-error-foreground);
}

</style>
        
        <link rel="stylesheet" href="https://cdn.jsdelivr.net/gh/Microsoft/vscode/extensions/markdown-language-features/media/markdown.css">
<link rel="stylesheet" href="https://cdn.jsdelivr.net/gh/Microsoft/vscode/extensions/markdown-language-features/media/highlight.css">
<style>
            body {
                font-family: -apple-system, BlinkMacSystemFont, 'Segoe WPC', 'Segoe UI', system-ui, 'Ubuntu', 'Droid Sans', sans-serif;
                font-size: 14px;
                line-height: 1.6;
            }
        </style>
        <style>
.task-list-item {
    list-style-type: none;
}

.task-list-item-checkbox {
    margin-left: -20px;
    vertical-align: middle;
    pointer-events: none;
}
</style>
        
    </head>
    <body class="vscode-body vscode-light">
        <h1 id="correlations-of-gene-family-sizes-vs-species-genome-characteristics">Correlations of gene family sizes vs. species genome characteristics</h1>
<p>We hypothesize that gene families proliferate more in genomes with a high repeat content, because some TEs can facilitate gene duplications. Therefore we calculate the linear regression in each significantly rapidly evolving orthogroup between gene family size and several genome characteristics such as genome size, overall repeat content, and repeat-category content.</p>
<h2 id="method">Method</h2>
<h3 id="phylogeneticlaly-independent-contrasts-pics">Phylogeneticlaly independent contrasts (PICs)</h3>
<p>I am using my implementation of the phylogenetically independent contrasts according to Felsenstein 1985 and the species tree generated by orthofinder, transformed to be ultrametric with the ape package in R. The code is here: <code>PhD_chapter1/src/compute_PIC.py</code>. Both explanatory and response variables are transformed.</p>
<ul>
<li>exclude <em>D. melanogaster</em> from the analysis</li>
</ul>
<h3 id="spearman-correlation">spearman correlation</h3>
<p>I calculate the spearman correlation with the scipy function. They are NOT log transformed.</p>
<h3 id="genome-characteristics">genome characteristics</h3>
<p>I test the gene family sizes in each orthogroup against several genome characteristics to find orthogroups that have significant correlations. Each group is FDR-corrected against multiple testing. The categories are these:</p>
<ul>
<li>Genome size (assembly size)</li>
<li>Overall genome repeat proportion</li>
<li>proportion of different repeat categories
<ul>
<li>Unclassified</li>
<li>Low_complexity</li>
<li>Simple_repeats</li>
<li>Rolling-circles</li>
<li>Retroelements</li>
<li>DNA_transposons</li>
<li>Small_RNA</li>
<li>Satellites</li>
</ul>
</li>
</ul>
<h2 id="results">Results</h2>
<p>There are 496 significantly rapidly evolving orthogroups in total, but each plot shows fewer because I excluded the ones with non-normal residuals as explained above. Each plot is the orthogroup size vs. the linear regression slope to also take into account that larger orthogroups might be more prone to correlations with repetitive elements.</p>
<p>No category shows significantly correlated orthogroups after multiple testing correction.</p>
<h3 id="genome-size">Genome size</h3>
<p>There are four significantly correlated orthogroups</p>
<ul>
<li>N0.HOG0002345: No gene family cluster. Snurportin, snRNA import into nucleus</li>
<li>N0.HOG0000761: Gene family cluster 20, transmembrane transport (olfactory). Multidrug-Resistance like Protein 1(MRP), see functional analysis for why that makes sense</li>
<li>N0.HOG0000296: Gene family cluster 31, chromatin organization and transcription regulation. uncharacterized protein</li>
<li>N0.HOG0000049: No gene family cluster. cag, DNA binding.</li>
</ul>
<p><img src="file:////Users/miltr339/work/PhD_code/PhD_chapter1/data/correlations/orthoDB_sig_OGs_vs_GS_coefficients_bh_corrected_PIC_vs_OG_size_sig_OGs_colors.svg" alt="correlation" width=30%></p>
<h3 id="repeat-content">Repeat content</h3>
<p>This is the general genome-wide repeat content with no consideration for how it is divided in the different repeat categories</p>
<p><img src="file:////Users/miltr339/work/PhD_code/PhD_chapter1/data/correlations/orthoDB_sig_OGs_vs_reps_coefficients_bh_corrected_PIC_vs_OG_size_sig_OGs_colors.svg" alt="correlation" width=30%></p>
<h3 id="repeat-categories">repeat categories</h3>
<p>There are two significant correlations:</p>
<ul>
<li><strong>Satellites:</strong> N0.HOG0000808. No gene family cluster. Kynurenine aminotransferase, kynurenic acid biosynthetic process</li>
<li><strong>Low complexity</strong> N0.HOG0000056. Gene family cluster 7 (odorant binding), Odorant-binding protein 56e(Obp56e), sexual reproduction; sensory perception of chemical stimulus; sensory perception of smell</li>
</ul>
<p><img src="file:////Users/miltr339/work/PhD_code/PhD_chapter1/data/correlations/orthoDB_sig_OGs_vs_Unclassified_coefficients_bh_corrected_PIC_vs_OG_size_sig_OGs_colors.svg" alt="correlation" width=29%>
   <img src="file:////Users/miltr339/work/PhD_code/PhD_chapter1/data/correlations/orthoDB_sig_OGs_vs_Low_complexity_coefficients_bh_corrected_PIC_vs_OG_size_sig_OGs_colors.svg" alt="correlation" width=30%>
   <img src="file:////Users/miltr339/work/PhD_code/PhD_chapter1/data/correlations/orthoDB_sig_OGs_vs_Simple_repeats_coefficients_bh_corrected_PIC_vs_OG_size_sig_OGs_colors.svg" alt="correlation" width=30%></p>
<p><img src="file:////Users/miltr339/work/PhD_code/PhD_chapter1/data/correlations/orthoDB_sig_OGs_vs_Rolling-circles_coefficients_bh_corrected_PIC_vs_OG_size_sig_OGs_colors.svg" alt="correlation" width=30%>
    <img src="file:////Users/miltr339/work/PhD_code/PhD_chapter1/data/correlations/orthoDB_sig_OGs_vs_Retroelements_coefficients_bh_corrected_PIC_vs_OG_size_sig_OGs_colors.svg" alt="correlation" width=30%>
    <img src="file:////Users/miltr339/work/PhD_code/PhD_chapter1/data/correlations/orthoDB_sig_OGs_vs_DNA_transposons_coefficients_bh_corrected_PIC_vs_OG_size_sig_OGs_colors.svg" alt="correlation" width=30%></p>
<p><img src="file:////Users/miltr339/work/PhD_code/PhD_chapter1/data/correlations/orthoDB_sig_OGs_vs_Small_RNA_coefficients_bh_corrected_PIC_vs_OG_size_sig_OGs_colors.svg" alt="correlation" width=30%>
    <img src="file:////Users/miltr339/work/PhD_code/PhD_chapter1/data/correlations/orthoDB_sig_OGs_vs_Satellites_coefficients_bh_corrected_PIC_vs_OG_size_sig_OGs_colors.svg" alt="correlation" width=30%></p>
        
        
    </body>
    </html>